### PR TITLE
General context simplify

### LINKED ISSUE #479: Move RNG and deinition to context and tidy
Move RNG and model definition to context and remove unnecessary interfaces.

### PR DESCRIPTION
**Pull request by jamesturner246**
Working towards #461. 

Resolves #479. 

There are just a few real changes here. More to follow.

* RNG is now in context only.
* Model definition is now in context only
* Remove unnecessary Demographic and Disease module interfaces
* General tidy up

### CODE REVIEW

File: src/HealthGPS/runtime_context.h [hunk 1]
```diff
@@ -100,9 +108,9 @@ class RuntimeContext {
 
   private:
     std::reference_wrapper<EventAggregator> event_bus_;
-    std::reference_wrapper<SimulationDefinition> definition_;
+    std::unique_ptr<SimulationDefinition> definition_;
```
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS/runtime_context.h:111]: That said, I *think* you might just be able to move `SimulationDefinition` objects themselves directly.... Unless you're using polymorphism?

**Review by alexdewar — APPROVED** [review]
The cleanups all seem sensible to me!

File: src/HealthGPS/random_algorithm.cpp [hunk 2]
```diff
@@ -6,7 +6,8 @@
 #include <stdexcept>
 
 namespace hgps {
-Random::Random(RandomBitGenerator &generator) : engine_{generator} {}
+
+void Random::seed(unsigned int seed) noexcept { engine_.seed(seed); }
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/random_algorithm.cpp:10]: Yeah, we don't, basically. `Random` just has more advanced RNG fns like normally distributed, etcetera. Something to be merged.

File: src/HealthGPS.Tests/Simulation.Test.cpp [hunk 3]
```diff
@@ -106,36 +106,36 @@ TEST(TestSimulation, RandomBitGeneratorCopy) {
 TEST(TestSimulation, RandomAlgorithmStandalone) {
     using namespace hgps;
 
-    auto rnd = MTRandom32(123456789);
-    auto rnd_gen = Random(rnd);
-    auto value = rnd_gen.next_double();
+    auto random = Random{};
+    random.seed(123456789);
+    auto value = random.next_double();
     ASSERT_GT(value, 0.0);
```
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS.Tests/Simulation.Test.cpp:112]: No idea :confused: I'll look into it.

File: src/HealthGPS/runtime_context.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS/runtime_context.h:111]: `SimulationDefinition` itself could be one for the chopping block. Just a wrapper struct around things that are arguably 'context'.

File: src/HealthGPS.Tests/Simulation.Test.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jamesturner246) [reply src/HealthGPS.Tests/Simulation.Test.cpp]: Alright a lot of tests are weird, and probably need looking at eventually. I've fixed these ones meanwhile.

File: src/HealthGPS/random_algorithm.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexdewar) [reply src/HealthGPS/random_algorithm.cpp:10]: Hmm that does sound useful actually. Although it still looks a bit crufty

**Pull request merged**